libp2p/peerstore.nim

```diff
@@ -33,33 +32,86 @@ type
     book*: Table[PeerID, HashSet[MultiAddress]]
     addrChange: AddrChangeHandler
   
+  ProtoBook* = object
+    book*: Table[PeerID, HashSet[string]]
+    protoChange: ProtoChangeHandler
+  
+  KeyBook* = object
+    book*: Table[PeerID, PublicKey]
+    keyChange: KeyChangeHandler
+  
+  MetadataBook* = object
+    book*: Table[PeerID, Table[string, seq[byte]]]
+    metadataChange: MetadataChangeHandler
+  
   ####################
   # Peer store types #
   ####################
 
   PeerStore* = ref object of RootObj
     addressBook*: AddressBook
-    listeners: seq[EventListener]
+    protoBook*: ProtoBook
+    keyBook*: KeyBook
+    metadataBook*: MetadataBook
+    addrChangeHandlers*: seq[AddrChangeHandler]
```

Comment:
Hmm, wouldn't it make more sense to move the change handlers into the specific `*Book` types? For example, I'm not sure what the functionality of `KeyBook.keyChange` is and how  it relates to `keyChangeHandlers`? Wouldn't it make more sense to move  `keyChangeHandlers` under `KeyBook` and get rid of `keyChange`, or am I missing something?

Reply:
I agree. :) Improvement in ff7d081

The idea was to create some level of indirection between the multiple _public_ handlers that different clients can register on the Peer Store and the single, _private_ handler that the peer store itself registers per *Book when it initialises those books. Each book then only notifies its encompassing Peer Store of changes and the Peer Store in turn notifies its multiple registered clients (perhaps enforcing optional policy). Currently this serves no more than a conceptual purpose, so it will be cleaner to just move the handlers to the individual books. Thanks!

Reply:
Yeah, I'm not sure if the "proxy" handler makes sense there - but I understand the idea behind it now. 

My perspective on the `PeerStore` is a wrapper that exposes specific `register*Handler` methods to each underlying `*Book`. I think it's cleaner and simpler to reason about, but I'm willing to reconsider it when we might require this functionality.